metadata/src/main/java/org/apache/kafka/controller/DelegationTokenControlManager.java

```diff
@@ -0,0 +1,359 @@
+        for (CreatableRenewers renewer : requestData.renewers()) {
+            if (renewer.principalType().equals(KafkaPrincipal.USER_TYPE)) {
+                renewers.add(new KafkaPrincipal(renewer.principalType(), renewer.principalName()));
+            } else {
+                return ControllerResult.atomicOf(records, responseData.setErrorCode(INVALID_PRINCIPAL_TYPE.code()));
+            }
+        }
+
+        byte[] hmac;
+        try {
+            hmac = createHmac(tokenId);
+        } catch (Throwable e) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(ApiError.fromThrowable(e).error().code()));
+        }
+
+        TokenInformation newTokenInformation = new TokenInformation(tokenId, owner,
+            context.principal(), renewers, now, maxTimestamp, expiryTimestamp);
+
+        DelegationTokenData newDelegationTokenData = new DelegationTokenData(newTokenInformation);
+
+        responseData
+                .setErrorCode(NONE.code())
+                .setIssueTimestampMs(now)
+                .setExpiryTimestampMs(expiryTimestamp)
+                .setMaxTimestampMs(maxTimestamp)
+                .setTokenId(tokenId)
+                .setHmac(hmac);
+
+        records.add(new ApiMessageAndVersion(newDelegationTokenData.toRecord(), (short) 0));
+        return ControllerResult.atomicOf(records, responseData);
+    }
+
+    public ControllerResult<RenewDelegationTokenResponseData> renewDelegationToken(
+        ControllerRequestContext context,
+        RenewDelegationTokenRequestData requestData,
+        MetadataVersion metadataVersion
+    ) {
+        long now = time.milliseconds();
+
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+        RenewDelegationTokenResponseData responseData = new RenewDelegationTokenResponseData();
+
+        TokenInformation myTokenInformation = getToken(requestData.hmac());
+
+        if (myTokenInformation == null) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_NOT_FOUND.code()));
+        }
+
+        if (myTokenInformation.maxTimestamp() < now || myTokenInformation.expiryTimestamp() < now) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_EXPIRED.code()));
+        }
+
+        if (!allowedToRenew(myTokenInformation, context.principal())) {
+            return ControllerResult.atomicOf(records, responseData.setErrorCode(DELEGATION_TOKEN_OWNER_MISMATCH.code()));
+        }
+
+        long renewLifeTime = tokenDefaultRenewLifetime;
+        if (requestData.renewPeriodMs() > 0) {
+            renewLifeTime = Math.min(renewLifeTime, requestData.renewPeriodMs());
+        }
+        long renewTimeStamp = now + renewLifeTime;
+        long expiryTimestamp = Math.min(myTokenInformation.maxTimestamp(), renewTimeStamp);
+
+        myTokenInformation.setExpiryTimestamp(expiryTimestamp);
```

Comment:
This is updating the record directly within the cache, but it is okay since this is the only thread accessing that cache -- is that correct?  Might be worth a comment to this effect in `getToken()` and `sweepExpiredDelegationTokens()` where we use the token cache.

How does the delegation token cache get updated in other nodes?  Shouldn't this updating of the token cache happen when the written record gets replayed, as opposed to here?

Reply:
I thought `myTokenInformation` is a copy of what is in the cache not a reference to it.

The copy is updated and and record is created using the copy. The replay of the record on brokers and kcontrollers updates the cache on each node.

Reply:
Fixed